{
  "active_lock_reason": null,
  "assignee": null,
  "assignees": [],
  "author_association": "NONE",
  "body": "**Description**\r\n\r\n<!--\r\nBriefly describe the problem you are having in a few paragraphs.\r\n-->\r\n\r\nFor v2 shim, during containerd start up, if there are left-overs tasks in the \"task\" folder, containerd would launch the shim with \"delete\" argument to do the clean up. But at that time, the containerd's ttrpc server isn't up and running, so the shim would exit with error as it cannot connect to the ttrpc server. This could result in the tasks in the `io.containerd.runtime.v2.task` folder can never be cleaned up? \r\n\r\n**Steps to reproduce the issue:**\r\n1. start containerd, and then use ctr to start a daemon task. For example:\r\n\r\n```\r\nctr run --runtime io.containerd.runc.v2 -d docker.io/library/nginx:latest mynginx\r\n```\r\n\r\n2. kill containerd first and kill the shim process. Or we can simple shutdown the host machine and reboot it.\r\n3. start containerd again, we will observer it tries to do clean up for the dead shim, but hit errors.\r\n\r\n```\r\nINFO[2020-01-19T02:10:57.604741661-05:00] loading plugin \"io.containerd.runtime.v2.task\"...  type=io.containerd.runtime.v2\r\nDEBU[2020-01-19T02:10:57.604922052-05:00] loading tasks in namespace                    namespace=default\r\nWARN[2020-01-19T02:10:57.605176979-05:00] cleaning up after shim disconnected           id=mynginx namespace=default\r\nINFO[2020-01-19T02:10:57.605215784-05:00] cleaning up dead shim                        \r\nWARN[2020-01-19T02:10:57.609013918-05:00] failed to clean up after shim disconnected    error=\"io.containerd.runc.v2: failed to connect: dial unix /run/containerd/containerd.sock.ttrpc: connect: connection refused\\n: exit status 1\" id=mynginx namespace=default\r\n```\r\n\r\nAnd there is no way we can clean up the task folder afterwards. Even if we use `ctr container delete` command to remove the container, we can no longer create container with the same name:\r\n\r\n```\r\n# ctr container rm mynginx\r\n# ctr run --runtime io.containerd.runc.v2 -d docker.io/library/nginx:latest mynginx\r\nctr: mkdir /run/containerd/io.containerd.runtime.v2.task/default/mynginx: file exists: unknown\r\n```\r\n\r\n**Describe the results you received:**\r\n\r\n```\r\nWARN[2020-01-19T02:10:57.609013918-05:00] failed to clean up after shim disconnected    error=\"io.containerd.runc.v2: failed to connect: dial unix /run/containerd/containerd.sock.ttrpc: connect: connection refused\\n: exit status 1\" id=mynginx namespace=default\r\n```\r\n\r\n**Describe the results you expected:**\r\n\r\nThe clean up should finish without error\r\n\r\n**Output of `containerd --version`:**\r\n\r\n```\r\ncontainerd github.com/containerd/containerd v1.3.0-279-g592a617f3 592a617f3d6e3c6aa0f1894c3edfe9a9cf890600\r\n```\r\n\r\n**Any other relevant information:**\r\n\r\nSimilar issue was discussed before ttrpc is introduced:\r\nhttps://github.com/containerd/containerd/pull/3204#issuecomment-482012201\r\n\r\nSeems this change re-introduce this issue:\r\nhttps://github.com/containerd/containerd/commit/7b06c9a1ce2b2d01bf3c6b13f8416db0ec4b0eca",
  "closed_at": "2020-03-14T22:19:49Z",
  "closed_by": {
    "avatar_url": "https://avatars2.githubusercontent.com/u/865334?v=4",
    "events_url": "https://api.github.com/users/mxpv/events{/privacy}",
    "followers_url": "https://api.github.com/users/mxpv/followers",
    "following_url": "https://api.github.com/users/mxpv/following{/other_user}",
    "gists_url": "https://api.github.com/users/mxpv/gists{/gist_id}",
    "gravatar_id": "",
    "html_url": "https://github.com/mxpv",
    "id": 865334,
    "login": "mxpv",
    "node_id": "MDQ6VXNlcjg2NTMzNA==",
    "organizations_url": "https://api.github.com/users/mxpv/orgs",
    "received_events_url": "https://api.github.com/users/mxpv/received_events",
    "repos_url": "https://api.github.com/users/mxpv/repos",
    "site_admin": false,
    "starred_url": "https://api.github.com/users/mxpv/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/mxpv/subscriptions",
    "type": "User",
    "url": "https://api.github.com/users/mxpv"
  },
  "comments": 1,
  "comments_url": "https://api.github.com/repos/containerd/containerd/issues/3971/comments",
  "created_at": "2020-01-19T07:19:29Z",
  "events_url": "https://api.github.com/repos/containerd/containerd/issues/3971/events",
  "html_url": "https://github.com/containerd/containerd/issues/3971",
  "id": 551879203,
  "labels": [
    {
      "color": "FF8C94",
      "default": false,
      "description": null,
      "id": 347599646,
      "name": "kind/bug",
      "node_id": "MDU6TGFiZWwzNDc1OTk2NDY=",
      "url": "https://api.github.com/repos/containerd/containerd/labels/kind/bug"
    }
  ],
  "labels_url": "https://api.github.com/repos/containerd/containerd/issues/3971/labels{/name}",
  "locked": false,
  "milestone": null,
  "node_id": "MDU6SXNzdWU1NTE4NzkyMDM=",
  "number": 3971,
  "performed_via_github_app": null,
  "repository_url": "https://api.github.com/repos/containerd/containerd",
  "state": "closed",
  "title": "containerd cannot properly do \"clean-up\" with shim process during start up?",
  "updated_at": "2020-03-14T22:19:49Z",
  "url": "https://api.github.com/repos/containerd/containerd/issues/3971",
  "user": {
    "avatar_url": "https://avatars0.githubusercontent.com/u/13135289?v=4",
    "events_url": "https://api.github.com/users/wswfc/events{/privacy}",
    "followers_url": "https://api.github.com/users/wswfc/followers",
    "following_url": "https://api.github.com/users/wswfc/following{/other_user}",
    "gists_url": "https://api.github.com/users/wswfc/gists{/gist_id}",
    "gravatar_id": "",
    "html_url": "https://github.com/wswfc",
    "id": 13135289,
    "login": "wswfc",
    "node_id": "MDQ6VXNlcjEzMTM1Mjg5",
    "organizations_url": "https://api.github.com/users/wswfc/orgs",
    "received_events_url": "https://api.github.com/users/wswfc/received_events",
    "repos_url": "https://api.github.com/users/wswfc/repos",
    "site_admin": false,
    "starred_url": "https://api.github.com/users/wswfc/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/wswfc/subscriptions",
    "type": "User",
    "url": "https://api.github.com/users/wswfc"
  }
}